### PR TITLE
chore: param validation in genesis

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

adds param validation in genesis validation 

Closes: #474
Closes: #476 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/genesis.go [hunk 1]
```diff
@@ -60,6 +61,78 @@ func (s *GenesisState) Validate() error {
 		return err
 	}
 
+	// run each params validation method
+	if err := validateParams(s.Params); err != nil {
```
Comment (aleem1314) [inline x/ecocredit/genesis.go:65]: How about `s.Params.Validate()` ?

Comment (technicallyty) [reply x/ecocredit/genesis.go]: yeah that looks much cleaner -> https://github.com/regen-network/regen-ledger/pull/511/commits/9cd97542fa296479d1c1fe356240b09934819430

File: x/ecocredit/genesis.go [hunk 2]
```diff
@@ -60,9 +61,61 @@ func (s *GenesisState) Validate() error {
...
+		// check that the other fields are equal
+		if !areCreditTypesEqual(cType, *cInfo.CreditType) {
+			return fmt.Errorf("credit type %s does not match param type %s", cInfo.CreditType.Name, cType.Name)
+		}
+	}
+	return nil
+}
+
+// checks that fields are equal between two credits.
+func areCreditTypesEqual(t1, t2 CreditType) bool {
```
Comment (ryanchristo) [inline x/ecocredit/genesis.go:103]: Might be nice to use a switch statement here.
```
switch {
case t1.Abbreviation != t2.Abbreviation:
case t1.Name != t2.Name:
case t1.Precision != t2.Precision:
case t1.Unit != t2.Unit:
    return false
default:
    return true
}
```

File: x/ecocredit/genesis.go [hunk 3]
```diff
@@ -60,9 +61,61 @@ func (s *GenesisState) Validate() error {
...
+	// convert to map for easier lookups
+	for _, cType := range creditTypes {
+		typeMap[cType.Abbreviation] = *cType
+	}
+
+	for _, cInfo := range classInfos {
+		// fetch param via abbreviation
+		cType, ok := typeMap[cInfo.CreditType.Abbreviation]
+
+		// if its not found, its an invalid credit type
```
Comment (ryanchristo) [inline x/ecocredit/genesis.go:89]:
```suggestion
		// if it's not found, it's an invalid credit type
```

File: x/ecocredit/genesis.go [hunk 4]
```diff
@@ -60,9 +61,61 @@ func (s *GenesisState) Validate() error {
...
+		if !areCreditTypesEqual(cType, *cInfo.CreditType) {
+			return fmt.Errorf("credit type %s does not match param type %s", cInfo.CreditType.Name, cType.Name)
+		}
+	}
+	return nil
+}
+
+// checks that fields are equal between two credits.
+func areCreditTypesEqual(t1, t2 CreditType) bool {
+	if t1.Abbreviation != t2.Abbreviation {
```
Comment (ryanchristo) [inline x/ecocredit/genesis.go:104]: Is this check necessary since we are looking up the credit type by abbreviation?

**Review by ryanchristo — COMMENTED** [review]
Looking good. It would be nice to add a couple more tests to improve code coverage. A few new lines are not covered in `areCreditTypesEqual` and one equals check might not be necessary.

File: x/ecocredit/genesis.go (hunk 2 above)
Comment (ruhatch) [reply x/ecocredit/genesis.go:103]: Out of interest, why doesn't `t1 == t2` work? Could you document that in this function?

Comment (technicallyty) [reply x/ecocredit/genesis.go]: oops, looks like you can just do `t1 == t2` updated here https://github.com/regen-network/regen-ledger/pull/511/commits/79cbd50fdd4ff8c8666f1f4d7d75a2852b0c3613#diff-e57c839aca633b049a6d8f6706c7b5c53322c1e7cc17157178da38461789043cR95

**Comment by technicallyty** [comment]
noticed some docs still mention credit designer when i was fixing merge errors. do we have an issue to track docs updating for this? @ryanchristo @ruhatch

File: x/ecocredit/genesis.go [hunk 5]
```diff
@@ -60,6 +61,41 @@ func (s *GenesisState) Validate() error {
...
+		typeMap[cType.Abbreviation] = *cType
+	}
+
+	for _, cInfo := range classInfos {
+		// fetch param via abbreviation
+		cType, ok := typeMap[cInfo.CreditType.Abbreviation]
+
+		// if it's not found, its an invalid credit type
+		if !ok {
+			return fmt.Errorf("unknown credit type abbreviation: %s", cInfo.CreditType.Abbreviation)
```
Comment (blushi) [inline x/ecocredit/genesis.go:91]: what about using `sdkerrors.ErrInvalidType.Wrapf("%s is not a valid credit type", creditTypeName)` instead, which is used in `serverImpl.getCreditType` to keep error msgs consistent?

File: x/ecocredit/params.go [hunk 6]
```diff
@@ -36,6 +36,27 @@ func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
 	}
 }
 
+// Validate will run each param field's validate method
+func (p Params) Validate() error {
+	if err := validateCreditTypes(p.CreditTypes); err != nil {
+		return err
+	}
+
+	if err := validateAllowedClassCreators(p.AllowedClassCreators); err != nil {
```
Comment (blushi) [inline x/ecocredit/params.go:45]: Although those functions are already tested independently, it'd be good still to add tests for the err cases within `Params.Validate` IMO

**Review by blushi — COMMENTED** [review]
Looks good overall, just some nits related to returned errors and test coverage

File: x/ecocredit/genesis.go (hunk 5 above)
Comment (technicallyty) [reply x/ecocredit/genesis.go:91]: now that i think about it - perhaps ErrNotFound would be better here

File: x/ecocredit/params.go (hunk 6 above)
Comment (technicallyty) [reply x/ecocredit/params.go:45]: ok, added some more cases here https://github.com/regen-network/regen-ledger/pull/511/commits/0201ea829e05c39a1e811bb939bd415ddf9848fa

File: x/ecocredit/genesis.go (hunk 5 above)
Comment (blushi) [reply x/ecocredit/genesis.go]: makes sense

**Review by ruhatch — APPROVED** [review]
Looks good!

**Review by ryanchristo — APPROVED** [review]
Looks good to me. Nice work!

**Review by blushi — APPROVED** [review]
utACK

**Review by atheeshp — APPROVED** [review]

**Comment by technicallyty** [comment]
could someone merge this in? i don't have authorization @ruhatch @ryanchristo ?

**Pull request merged**